internal/speakerlist/speakerlist.go

```diff
@@ -0,0 +1,301 @@
+		logger.Log("op", "startup", "error", err, "msg", "failed to create memberlist")
+		return nil, err
+	}
+	return &sl, nil
+}
+
+// Start starts the needed goroutines
+func (sl *SpeakerList) Start(client *k8s.Client) error {
+	sl.client = client
+
+	if sl.mList == nil {
+		return nil
+	}
+
+	go sl.memberlistWatchEvents()
+	go sl.mlJoinLoop()
+
+	return nil
+}
+
+// updateMlIPs runs forever updating the sl.mlSpeakersIPs slice with the
+// IPs of all the speaker pods. As this queries the API server, it waits at
+// least 5 minutes between queries, to avoid self-induced API server DoS (should
+// be trated with care).
+func (sl *SpeakerList) speakersIPsLoop() {
+	for {
+		// Create the ticker at the start of each loop, so we wait at
+		// least the ticker interval before the ticker case is true.
+		// If we create it outside of the loop, it will tick every 5
+		// minutes and if the blocking call happens to take longer, the
+		// next tick will be really soon and we may overload the API
+		// server.
+		ticker := time.NewTicker(5 * time.Minute)
+
+		select {
+		case <-sl.stopCh:
+			return
+		case <-ticker.C:
+			// This blocks until the apiserver responds.
+			iplist, err := sl.mlSpeakers()
+			if err != nil {
+				sl.l.Log("op", "memberDiscovery", "error", err, "msg", "failed to get pod IPs")
+				continue
+			}
+
+			sl.mlMux.Lock()
+			sl.mlSpeakersIPs = iplist
+			sl.mlMux.Unlock()
+		}
+	}
+}
+
+func (sl *SpeakerList) mlSpeakers() ([]string, error) {
+	// This call blocks until we get a response of the API server.
+	// In the client-go version we are using, there is no way to use a
+	// context or timeout.
+	// TODO: When updating client-go, this code can be simplified by using a
+	// context in the following way:
+	// If we use a context, we can specify a timeout and there is no need to
+	// have the speakersIPsLoop() running in the background. This can just
+	// be called from mlJoin() with a context (timeout) and if doesn't
+	// answer, use sl.mlSpeakersIPs.  If it doesn't time out, just update
+	// sl.mlSpeakersIPs with the answer and continue to do the join.
+	iplist, err := sl.client.GetPodsIPs(sl.namespace, sl.labels)
+	if err != nil {
+		return nil, err
+	}
+
+	return iplist, nil
+}
+
+func (sl *SpeakerList) mlJoinLoop() {
+	iplist, err := sl.mlSpeakers()
+	if err != nil {
+		sl.l.Log("op", "memberDiscovery", "error", err, "msg", "failed to get pod IPs")
+		iplist = nil
+	}
+
+	// Initialize sl.mlSpeakersIPs.
+	sl.mlMux.Lock()
+	sl.mlSpeakersIPs = iplist
+	sl.mlMux.Unlock()
+
+	// Update mlSpeakersIPs in the background.
+	go sl.speakersIPsLoop()
+
+	// Every one minute, try to rejoin.
+	// This will join nodes that leave the cluster only for a few seconds every 1
+	// minute, while discovering new IPs (updated by sl.speakersIPsLoop()) might
+	// take a while longer.
+	ticker := time.NewTicker(1 * time.Minute)
+
+	for {
+		select {
+		case <-sl.stopCh:
+			return
+		case <-ticker.C:
+			sl.mlJoin()
+		case <-sl.mlJoinCh:
```

Comment:
Yeap, the list is stale. But if only one IP is from a speaker pod, it will join the memberlist cluster successfully and the list will be updated in the background. 

That would be improved, of course, if we rely on some watching mechanism.

Reply:
Marked as resolved, as it is a known trade-off. Will be improved in future minor releases probably using a watch mechanism.